_posts/2019-12-12-report-ogp-engine-development.md

```diff
@@ -0,0 +1,129 @@
+
+Every object in the world will have an associated material, and objects that have a mass lower than an agent's carrying capacity can be held. Tools are incidentally just the name given to the items that are held by agents - functionally there is no difference between a tool and an ordinary object in the world.
+
+`Using a rock against a large plant`
+![Using a rock against a large plant](/assets/2019-12-12-report-ogp-engine-development/tools.jpg)
+
+## Mesh Slicing
+
+As an attempt to further proceduralize tool interactions, work was made to develop a mesh slicing algorithm that can be used as a tool action. This mechanic is especially useful for cutting down trees and subdividing large resources into carryable sections.
+
+`Slicing a complex mesh`
+![Slicing a complex mesh](/assets/2019-12-12-report-ogp-engine-development/slice2.png)
+![Slicing a complex mesh](/assets/2019-12-12-report-ogp-engine-development/slice1.png)
+
+`Slicing a large boulder. Texture coordinates have not been calculated, leaving a monocolor surface`
+![Slicing a complex mesh](/assets/2019-12-12-report-ogp-engine-development/slice3.jpg)
+## Code Cleanup
+
+### System Registration
+Systems are this engine's bread and butter, so it makes sense for these areas to be concise and efficient. With the help of C++ Macros, it was possible to reduce code duplication in some key areas.
+
+`Before`
+![Before](/assets/2019-12-12-report-ogp-engine-development/systems_messy.png)
+
+`After`
+
+![After](/assets/2019-12-12-report-ogp-engine-development/systems_clean.png)
+
+### Data Access
+
+As systems access entity components, synchronization mechanisms must be im place to ensure thread safety when locking these resources for use. The primary locking pattern has been enhanced to include a set of regions - regions are large (64m x 64m) square spatial subdivisions of the world that do not overlap. In a nutshell, there is a query signature composed of a Read mask, Write mask, and Region set. When a thread executes a task, it grabs a lock on this signature; if there are read/write conflicts, the region set is checked for intersection to determine if there really is a conflict. When a signature is released, all other blocked threads are notified and attempt to grab a lock on their signature. Read and write masks are 64 bit integers, where each bit is a flag for a specific component type, be it Position, Movement, Plant, etc.
+The region set is an unordered set of integer region IDs. 
+
+`Example usage:`
+![Example usage](/assets/2019-12-12-report-ogp-engine-development/data_access.png)
+
+The advantage of adding this spatial component to the lock signature is increased parallelism. Due to the localized nature of all world systems, there is never a case where a system update requires a query that spans more than one region away. To illustrate this concept, consider the following diagrams.
```

Comment:
```suggestion
The advantage of adding this spatial component to the lock signature is increased parallelism. Because of the localized nature of all world systems, there is never a case where a system update requires a query that spans more than one region away. To illustrate this concept, consider the following diagrams.
```